src/platform/PICO/bus_quadspi_pico.c

```diff
@@ -0,0 +1,371 @@
+{
+    UNUSED(isRead);
+    // Build tx: [cmd][addr bytes MSB..LSB][dummy]
+    uint8_t hdr[1 + 4 + 1];
+    int idx = 0;
+    hdr[idx++] = instruction;
+    for (int i = (addressSize/8) - 1; i >= 0; i--) {
+        hdr[idx++] = (address >> (i*8)) & 0xFF;
+    }
+    // Dummy cycles rounded to bytes
+    while (dummyCycles >= QSPI_DUMMY_BITS_PER_BYTE) {
+        hdr[idx++] = 0x00;
+        dummyCycles -= QSPI_DUMMY_BITS_PER_BYTE;
+    }
+    pico_qspi_enter_cmd_mode();
+    // Send header
+    for (int i = 0; i < idx; i++) {
+        uint8_t b = hdr[i];
+        flash_do_cmd(&b, &b, 1);
+    }
+    return true;
+}
+
+bool quadSpiReceiveWithAddress1LINE(QUADSPI_TypeDef *instance, uint8_t instruction, uint8_t dummyCycles,
+                                    uint32_t address, uint8_t addressSize, uint8_t *in, int length)
+{
+    UNUSED(instance);
+    // Build header and perform in one CS window
+    uint8_t hdr[1 + 4 + 1];
+    int idx = 0;
+    hdr[idx++] = instruction;
+    for (int i = (addressSize/8) - 1; i >= 0; i--) hdr[idx++] = (address >> (i*8)) & 0xFF;
+    while (dummyCycles >= QSPI_DUMMY_BITS_PER_BYTE) {
+        hdr[idx++] = 0x00;
+        dummyCycles -= QSPI_DUMMY_BITS_PER_BYTE;
+    }
+    pico_qspi_enter_cmd_mode();
+    qmi_direct_enable();
+    qmi_cs1_assert(true);
+    if (!qmi_direct_io(hdr, NULL, (size_t)idx, QSPI_TIMEOUT_MS)) {
+        qmi_timeout_cleanup();
+        return false;
+    }
+    // Read payload
+    for (int i = 0; i < length; i++) {
+        uint8_t tx = 0x00;
+        if (!qmi_direct_io(&tx, &in[i], 1, QSPI_TIMEOUT_MS)) {
+            qmi_timeout_cleanup();
+            return false;
+        }
+    }
+    qmi_cs1_assert(false);
+    qmi_direct_disable();
+    pico_qspi_exit_cmd_mode();
+    return true;
+}
+
+bool quadSpiTransmitWithAddress1LINE(QUADSPI_TypeDef *instance, uint8_t instruction, uint8_t dummyCycles,
+                                     uint32_t address, uint8_t addressSize, const uint8_t *out, int length)
+{
+    UNUSED(instance);
+    uint8_t hdr[1 + 4 + 1];
+    int idx = 0;
+    hdr[idx++] = instruction;
+    for (int i = (addressSize/8) - 1; i >= 0; i--) hdr[idx++] = (address >> (i*8)) & 0xFF;
+    while (dummyCycles >= QSPI_DUMMY_BITS_PER_BYTE) {
+        hdr[idx++] = 0x00;
+        dummyCycles -= QSPI_DUMMY_BITS_PER_BYTE;
+    }
+    pico_qspi_enter_cmd_mode();
+    qmi_direct_enable();
+    qmi_cs1_assert(true);
+    if (!qmi_direct_io(hdr, NULL, (size_t)idx, QSPI_TIMEOUT_MS)) {
+        qmi_timeout_cleanup();
+        return false;
+    }
+    if (out && length > 0) {
+        if (!qmi_direct_io(out, NULL, (size_t)length, QSPI_TIMEOUT_MS)) {
+            qmi_timeout_cleanup();
+            return false;
+        }
+    }
+    qmi_cs1_assert(false);
+    qmi_direct_disable();
+    pico_qspi_exit_cmd_mode();
+    return true;
+}
+
+// 4LINE operations are not truly available via ROM command mode on QMI; emulate using 1LINE for now.
```

Comment:
(isn't it better to tell driver that these functions are not available / emulate on upper level?)

Reply:
Maybe. To be honest I'm still trying to get my head round this stuff, aiming to move to DMA operation which may impact this.